Document the current process for merging from FreeBSD (#1137)
CTSRD-CHERI/cheribsd: CHERI-UPDATING.md

We typically merge to [master] from [dev] at stable points at least a week
apart.

### The merge process
CheriBSD is extremely diverged from FreeBSD with nearly two
thousand changed files. As such, there are many opportunities for
a merge to conflict and worse for a merge to not conflict, but
be broken. It is critical that we be able to bisect the merged
changes. To support this we merge changes one commit at a time
from the _first-parent_ history of FreeBSD's `main` using the
[mergify](https://github.com/brooksdavis/mergify) script. This makes it
straightforward to discover which change caused a problem.

This section describes the merge process in detail.

#### Setup
The merger must add an appropriate FreeBSD remote to their setup. This
example uses the official FreeBSD mirror, a mirror on GitHub or the like
will also work fine.
```
$ git remote add upstream https://git.FreeBSD.org/src.git
```
You will also want to enable `git revere` to simplify handling the same
conflicting merges over again. To do so globally, run:
```
$ git config --global rerere.enabled true
```
Disable rename detection. This sometimes leads to awkward merges but

> rename detection takes **forever** on the CheriBSD repo.

Reviewer: Want to give the command?

Reviewer 2: My guess is `git config merge.renames false`

FWIW, I did this as a local config in the checkout vs doing it globally.

```
Then identify the head of the set of commits to merge. Because FreeBSD
does not enforce date linearity this is a bit annoying. The default
`Date:` will generally be somewhere in the past relative to the actual
push unless the committer reset the date immediately before with
`git rebase --ignore-date` or `git commit --amend --date=now`. As
such we want to look at the commit date which is a better (but still
sub-optimal) proxy for the time the change was pushed to the FreeBSD
tree. We currently do this by examining the log and looking for the
first `CommitDate: ` line mentioning the day we wish to merge. To see
only eligible commits and to cause the commit date to be displayed we
use:
```
$ git log --first-parent --format=fuller upstream/main
```
We can then use the pager's search function to search for the
appropriate date (e.g.. `Commit.*Fri Feb 12`). It's a good idea to check
a few commits up or down to make sure the commit isn't at an obviously
bad point. Then we create a tag for the commit:
```
$ git tag -m "FreeBSD main DD Month YYYY" freebsd-main-<YYYYMMDD> <hash>
```

#### Performing the merge
First, create a merge branch from an update to date `dev` checkout:
```

> $ git checkout -b merge-freebsd-<YYYYMMDD>

Reviewer: I think you can use
```suggestion
$ git checkout -b merge-freebsd-<YYYYMMDD> dev
```
to help ensure that you start from the right place, regardless of the current checkout?

change or resolving the conflict, continue with:
```
mergify continue
```

For ease of conflict resolution, `mergify resolve` opens each file
containing conflict files in `EDITOR` in turn. `mergify autoadd` adds
all conflicted files that no longer contain conflict markers to the
index. It does not add files that were not in the original merge, they
need to be added with `git add`.

Sometimes a change has already been applied to the tree or is a change
requiring porting that is reverted later on. In that case, `mergify
skip` can be used. It restores all modified files to the version in HEAD
and then continues.

#### Rolling back
Over the course of the merge process, it is not uncommon to discover
that a previously merged change requires additional changes. Ideally
we'd make a fixup commit and then rebase, but `git rebase` is useless
with merge commits. Instead, the general process is to stop the merge,
use `git reset` to rewind to the broken commit, fix it up, and restart
the merge. The process usually looks like
```
$ mergify abort
$ git reset --hard <bad merge>
...
$ git commit --amend
$ mergify start freebsd-main-<YYYYMMDD>
```

> Sometime the commit it self isn't the problem, but rather that it

Reviewer: ```suggestion
Sometimes the commit itself isn't the problem, but rather that it
```

enabled a feature than was broken with CHERI. When this happens,
resetting to the commit before the feature was enabled, fixing things,
and restarting can make sense.

Sometimes there are too many awkward changes (e.g. one that required
fixup in files outside the original change) and it's not worth
backtracking to fix things in the ideal location. In that case, make
sure to indicate that the fixup commit fixes an issue introduced in a
particular commit to aid future users of bisect.

#### Finishing up
Once the merge is complete and basic local testing is done, it's time to
update the `freebsd-main` branch and push the tag. (It's good to delay
to this point in hopes of keeping `freebsd-main` relatively in sync with
main.)
```
$ git push origin freebsd-main-<YYYYMMDD>:freebsd-main
$ git push origin freebsd-main-<YYYYMMDD>
```

> Now, push the main tree to create a pull request for kick off a full CI run.

Reviewer 2: ```suggestion
Now, push the main tree to create a pull request to kick off a full CI run.
```

we'd make a fixup commit and then rebase, but `git rebase` is useless
with merge commits. Instead, the general process is to stop the merge,
use `git reset` to rewind to the broken commit, fix it up, and restart
the merge. The process usually looks like
```
$ mergify abort
$ git reset --hard <bad merge>
...
$ git commit --amend
$ mergify start freebsd-main-<YYYYMMDD>
```
Sometime the commit itself isn't the problem, but rather that it
enabled a feature than was broken with CHERI. When this happens,
resetting to the commit before the feature was enabled, fixing things,
and restarting can make sense.

Sometimes there are too many awkward changes (e.g. one that required
fixup in files outside the original change) and it's not worth
backtracking to fix things in the ideal location. In that case, make
sure to indicate that the fixup commit fixes an issue introduced in a
particular commit to aid future users of bisect.

#### Finishing up
Once the merge is complete and basic local testing is done, it's time to
update the `freebsd-main` branch and push the tag. (It's good to delay
to this point in hopes of keeping `freebsd-main` relatively in sync with
main.)
```

> $ git push origin freebsd-main-<YYYYMMDD>:freebsd-main

Reviewer 2: Sadly, this command did not work and I had to resort to checking out `freebsd-main`, merging the change, and then pushing `freebsd-main`.  This seems like maybe a bug in GH?
```suggestion
$ git checkout freebsd-main
$ git merge freebsd-main-<YYYYMMDD>
$ git push
```

Reviewer: Was it complaining that `freebsd-main-YYYYMMDD` wasn't a fast-forward from `freebsd-main`?

Reviewer 2: No, it's some other error about "bad ref" or the like.  I've lost it in my scroll back.  I'll post it when I do the next merge.

Reviewer 2: Found the error in Slack scroll back:

```
> git push origin freebsd-main-20210305:freebsd-main
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To github.com:CTSRD-CHERI/cheribsd.git
 ! [remote rejected]           freebsd-main-20210305 -> freebsd-main (bad ref)
error: failed to push some refs to 'github.com:CTSRD-CHERI/cheribsd.git'
```

Reviewer 2: I think this is the only thing that might need fixing before merging this in?

Author: Yeah, I'm still confused as to why this doesn't work. We might just want to suggest a worktree.

Reviewer 2: I'm effectively using a worktree currently.

Author: I've restored the worktree based text from the original commit.

skip` can be used. It restores all modified files to the version in HEAD
and then continues.

#### Rolling back
Over the course of the merge process, it is not uncommon to discover
that a previously merged change requires additional changes. Ideally
we'd make a fixup commit and then rebase, but `git rebase` is useless
with merge commits. Instead, the general process is to stop the merge,
use `git reset` to rewind to the broken commit, fix it up, and restart
the merge. The process usually looks like
```
$ mergify abort
$ git reset --hard <bad merge>
...
$ git commit --amend
$ mergify start freebsd-main-<YYYYMMDD>
```
Sometime the commit itself isn't the problem, but rather that it
enabled a feature than was broken with CHERI. When this happens,
resetting to the commit before the feature was enabled, fixing things,
and restarting can make sense.

Sometimes there are too many awkward changes (e.g. one that required
fixup in files outside the original change) and it's not worth
backtracking to fix things in the ideal location. In that case, make
sure to indicate that the fixup commit fixes an issue introduced in a
particular commit to aid future users of bisect.

#### Finishing up

> Once the merge is complete and basic local testing is done, it's time to

Reviewer 2: One note I guess is that I've been deferring doing these steps until the merge PR completes successfully and push all three things (new tag, freebsd-main, dev) at once.
